File: src/main/java/com/rs/game/content/combat/CombatDefinitions.java
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//  Copyright (C) 2021 Trenton Kress
//  This file is part of project: Darkan
//
package com.rs.game.content.combat;

import java.util.Map;

import com.rs.cache.loaders.Bonus;
import com.rs.game.model.entity.player.Equipment;
import com.rs.game.model.entity.player.Player;
import com.rs.game.model.entity.player.managers.InterfaceManager.Sub;
import com.rs.lib.game.Item;

public final class CombatDefinitions {
	
	public enum Spellbook {
		MODERN(192),
		ANCIENT(193),
		LUNAR(430),
		DUNGEONEERING(950);
		
		private int interfaceId;
		
		private Spellbook(int interfaceId) {
			this.interfaceId = interfaceId;
		}
		
		public int getInterfaceId() {
			return interfaceId;
		}
	}

	private transient Player player;
	private transient boolean usingSpecialAttack;
	private transient int[] bonuses;
	private transient boolean dungSpellbook;

	private byte attackStyle;
	private byte specialAttackPercentage;
	private boolean autoRetaliate;
	private byte sortSpellBook;
	private boolean showCombatSpells;
	private boolean showSkillSpells;
	private boolean showMiscSpells;
	private boolean showTeleportSpells;
	private boolean defensiveCasting;
	private Spellbook spellbook = Spellbook.MODERN;
	private CombatSpell autoCast;

	public CombatSpell getSpell() {
		CombatSpell spell = player.getTempAttribs().getO("manualCastSpell");
		if (spell != null)
			return spell;
		return autoCast;
	}

	public boolean hasManualCastQueued() {
		return player.getTempAttribs().getO("manualCastSpell") != null;
	}

	public void setManualCastSpell(CombatSpell spell) {
		player.getTempAttribs().setO("manualCastSpell", spell);
	}

	public void clearManualCastSpell() {
		player.getTempAttribs().removeO("manualCastSpell");
	}

	public CombatSpell getAutoCast() {
		return autoCast;
	}

	public void resetSpells(boolean removeAutoSpell) {
		clearManualCastSpell();
		if (removeAutoSpell) {
			setAutoCastSpell(null);
			refreshAutoCastSpell();
		}
	}

	public void setAutoCastSpell(CombatSpell spell) {
		autoCast = spell;
		refreshAutoCastSpell();
	}

	public void refreshAutoCastSpell() {
		refreshAttackStyle();
		player.getVars().setVar(108, getSpellAutoCastConfigValue());
	}

	public int getSpellAutoCastConfigValue() {
		if (autoCast == null)
			return 0;
		if (dungSpellbook)
			switch (autoCast) {
			case WIND_STRIKE:
				return 103;
			case WATER_STRIKE:
				return 105;
			case EARTH_STRIKE:
				return 107;
			case FIRE_STRIKE:
				return 109;
			case WIND_BOLT:
				return 111;
			case WATER_BOLT:
				return 113;
			case EARTH_BOLT:
				return 115;
			case FIRE_BOLT:
				return 117;
			case WIND_BLAST:
				return 119;
			case WATER_BLAST:
				return 121;
			case EARTH_BLAST:
				return 123;
			case FIRE_BLAST:
				return 125;
			case WIND_WAVE:
				return 127;
			case WATER_WAVE:
				return 129;
			case EARTH_WAVE:
				return 131;
			case FIRE_WAVE:
				return 133;
			case WIND_SURGE:
				return 135;
			case WATER_SURGE:
				return 137;
			case EARTH_SURGE:
				return 139;
			case FIRE_SURGE:
				return 141;
			default:
				return 0;
			}
		if (spellbook == Spellbook.MODERN)
			switch (autoCast) {
			case WIND_STRIKE:
				return 3;
			case WATER_STRIKE:
				return 5;
			case EARTH_STRIKE:
				return 7;
			case FIRE_STRIKE:
				return 9;
			case WIND_BOLT:
				return 11;
			case WATER_BOLT:
				return 13;
			case EARTH_BOLT:
				return 15;
			case FIRE_BOLT:
				return 17;
			case CRUMBLE_UNDEAD:
				return 35;
			case WIND_BLAST:
				return 19;
			case WATER_BLAST:
				return 21;
			case EARTH_BLAST:
				return 23;
			case FIRE_BLAST:
				return 25;
			case IBAN_BLAST:
				return 45;
			case MAGIC_DART:
				return 37;
			case SARADOMIN_STRIKE:
				return 41;
			case CLAWS_OF_GUTHIX:
				return 39;
			case FLAMES_OF_ZAMORAK:
				return 43;
			case WIND_WAVE:
				return 27;
			case WATER_WAVE:
				return 29;
			case EARTH_WAVE:
				return 31;
			case FIRE_WAVE:
				return 33;
			case WIND_SURGE:
				return 47;
			case WATER_SURGE:
				return 49;
			case EARTH_SURGE:
				return 51;
			case FIRE_SURGE:
				return 53;
			case WIND_RUSH:
				return 143;
			case STORM_OF_ARMADYL:
				return 145;
			default:
				return 0;
			}
		else if (spellbook == Spellbook.ANCIENT)
			switch (autoCast) {
			case SMOKE_RUSH:
				return 63;
			case SHADOW_RUSH:
				return 65;
			case BLOOD_RUSH:
				return 67;
			case ICE_RUSH:
				return 69;
			case SMOKE_BURST:
				return 71;
			case SHADOW_BURST:
				return 73;
			case BLOOD_BURST:
				return 75;
			case ICE_BURST:
				return 77;
			case SMOKE_BLITZ:
				return 79;
			case SHADOW_BLITZ:
				return 81;
			case BLOOD_BLITZ:
				return 83;
			case ICE_BLITZ:
				return 85;
			case SMOKE_BARRAGE:
				return 87;
			case SHADOW_BARRAGE:
				return 89;
			case BLOOD_BARRAGE:
				return 91;
			case ICE_BARRAGE:
				return 93;
			case MIASMIC_RUSH:
				return 95;
			case MIASMIC_BURST:
				return 97;
			case MIASMIC_BLITZ:
				return 99;
			case MIASMIC_BARRAGE:
				return 101;
			default:
				return 0;
			}
		else
			return 0;
	}

	public CombatDefinitions() {
		specialAttackPercentage = 100;
		autoRetaliate = true;
		showCombatSpells = true;
		showSkillSpells = true;
		showMiscSpells = true;
		showTeleportSpells = true;
	}

	public void setSpellbook(Spellbook book) {
		if (book == Spellbook.DUNGEONEERING)
			dungSpellbook = true;
		else
			spellbook = book;
		refreshSpellbook();
		player.getInterfaceManager().sendSubDefault(Sub.TAB_MAGIC);
	}

	public Spellbook getSpellbook() {
		if (dungSpellbook)
			return Spellbook.DUNGEONEERING;
		return spellbook;
	}

	public void switchShowCombatSpells() {
		showCombatSpells = !showCombatSpells;
		refreshSpellbookSettings();
	}

	public void switchShowSkillSpells() {
		showSkillSpells = !showSkillSpells;
		refreshSpellbookSettings();
	}

	public void switchShowMiscSpells() {
		showMiscSpells = !showMiscSpells;
		refreshSpellbookSettings();
	}

	public void switchShowTeleportSkillSpells() {
		showTeleportSpells = !showTeleportSpells;
		refreshSpellbookSettings();
	}

	public void switchDefensiveCasting() {
		defensiveCasting = !defensiveCasting;
		refreshSpellbookSettings();
	}

	public void setSortSpellBook(int sortId) {
		sortSpellBook = (byte) sortId;
		refreshSpellbookSettings();
	}

	public boolean isDefensiveCasting() {
		return defensiveCasting;
	}

	public void refreshSpellbookSettings() {
		player.getVars().setVarBit(357, spellbook.ordinal());
		player.getVars().setVarBit(5822, sortSpellBook);
		player.getVars().setVarBit(5823, sortSpellBook);
		player.getVars().setVarBit(5824, sortSpellBook);
		player.getVars().setVarBit(7347, sortSpellBook);

		player.getVars().setVarBit(6459, showCombatSpells ? 0 : 1);
		player.getVars().setVarBit(6466, showCombatSpells ? 0 : 1);
		player.getVars().setVarBit(6463, showCombatSpells ? 0 : 1);
		player.getVars().setVarBit(7348, showCombatSpells ? 0 : 1);

		player.getVars().setVarBit(6460, showSkillSpells ? 0 : 1);
		player.getVars().setVarBit(7349, showSkillSpells ? 0 : 1);

		player.getVars().setVarBit(6461, showMiscSpells ? 0 : 1);
		player.getVars().setVarBit(6464, showMiscSpells ? 0 : 1);
		player.getVars().setVarBit(7350, showMiscSpells ? 0 : 1);

		player.getVars().setVarBit(6462, showTeleportSpells ? 0 : 1);
		player.getVars().setVarBit(6467, showTeleportSpells ? 0 : 1);
		player.getVars().setVarBit(6465, showTeleportSpells ? 0 : 1);
		player.getVars().setVarBit(7351, showTeleportSpells ? 0 : 1);

		player.getVars().setVarBit(2668, defensiveCasting ? 1 : 0);
	}

	public static final Bonus getMeleeDefenceBonus(Bonus attackType) {
		switch(attackType) {
		case STAB_ATT:
			return Bonus.STAB_DEF;
		case SLASH_ATT:
			return Bonus.SLASH_DEF;
		default:
			return Bonus.CRUSH_DEF;
		}
	}

	public void setPlayer(Player player) {
		this.player = player;
		bonuses = new int[18];
		if (spellbook == null)
			spellbook = Spellbook.MODERN;
	}

	public int getBonus(Bonus bonus) {
		return bonuses[bonus.ordinal()];
	}

	public void setBonus(Bonus bonus, int val) {
		bonuses[bonus.ordinal()] = val;
	}

	public void refreshBonuses() {
		bonuses = new int[18];
		for (Item item : player.getEquipment().getItemsCopy()) {
			if (item == null)
				continue;
			for (Bonus bonus : Bonus.values()) {
				if (bonus == Bonus.RANGE_STR && getBonus(Bonus.RANGE_STR) != 0)
					continue;
				this.bonuses[bonus.ordinal()] += Equipment.getBonus(player, item, bonus);
			}
		}
	}

	public void resetSpecialAttack() {
		drainSpec(0);
		specialAttackPercentage = 100;
		refreshSpecialAttackPercentage();
	}

	public void setSpecialAttack(int special) {
		drainSpec(0);
		specialAttackPercentage = (byte) special;
		refreshSpecialAttackPercentage();
	}

	public void restoreSpecialAttack() {
		if (player.getFamiliar() != null)
			player.getFamiliar().restoreSpecialAttack(15);
		if (specialAttackPercentage == 100)
			return;
		restoreSpecialAttack(10);
		if (specialAttackPercentage == 100 || specialAttackPercentage == 50)
			player.sendMessage("<col=00FF00>Your special attack energy is now " + specialAttackPercentage + "%.", true);
	}

	public void restoreSpecialAttack(int percentage) {
		if (specialAttackPercentage >= 100 || player.getInterfaceManager().containsScreenInter())
			return;
		specialAttackPercentage += specialAttackPercentage > (100 - percentage) ? 100 - specialAttackPercentage : percentage;
		refreshSpecialAttackPercentage();
	}

	public void init() {
		refreshUsingSpecialAttack();
		refreshSpecialAttackPercentage();
		refreshAutoRelatie();
		refreshAttackStyle();
		refreshSpellbook();
	}

	public void refreshSpellbook() {
		refreshSpellbookSettings();
		refreshAutoCastSpell();
		player.getVars().syncVarsToClient();
		player.getPackets().sendRunScriptBlank(2057);
	}

	public void checkAttackStyle() {
		if (autoCast == null)
			setAttackStyle(attackStyle);
	}

	public void setAttackStyle(int style) {
		Map<Integer, AttackStyle> styles = AttackStyle.getStyles(player.getEquipment().getWeaponId());
		if (style < 0)
			style = 0;
		for (int i = style;i >= 0;i--) {
			if (styles.get(i) != null) {
				style = i;
				break;
			}
		}
		if (style != attackStyle) {
			attackStyle = (byte) style;
			if (autoCast != null)
				resetSpells(true);
			else
				refreshAttackStyle();
		} else if (autoCast != null)
			resetSpells(true);
	}

	public void refreshAttackStyle() {
		player.getVars().setVar(43, autoCast != null ? 4 : attackStyle);
	}

	public void sendUnlockAttackStylesButtons() {
		for (int componentId = 7; componentId <= 10; componentId++)
			player.getPackets().setIFRightClickOps(884, componentId, -1, 0, 0);
	}

	public void switchUsingSpecialAttack() {
		usingSpecialAttack = !usingSpecialAttack;
		refreshUsingSpecialAttack();
	}

	public void drainSpec(int amount) {
		usingSpecialAttack = false;
		refreshUsingSpecialAttack();
		if (player.getNSV().getB("infSpecialAttack"))
			amount = 0;
		if (amount > 0) {
			specialAttackPercentage -= amount;
			refreshSpecialAttackPercentage();
		}
	}

	public boolean hasRingOfVigour() {
		return player.getEquipment().getRingId() == 19669;
	}

	public int getSpecialAttackPercentage() {
		return specialAttackPercentage;
	}

	public void refreshUsingSpecialAttack() {
		player.getVars().setVar(301, usingSpecialAttack ? 1 : 0);
	}

	public void refreshSpecialAttackPercentage() {
		player.getVars().setVar(300, specialAttackPercentage * 10);
	}

	public void switchAutoRetaliate() {
		autoRetaliate = !autoRetaliate;
		refreshAutoRelatie();
	}

	public void refreshAutoRelatie() {
		player.getVars().setVar(172, autoRetaliate ? 0 : 1);
	}

	public boolean isUsingSpecialAttack() {
		return usingSpecialAttack;
	}

	public AttackStyle getAttackStyle() {
		Map<Integer, AttackStyle> styles = AttackStyle.getStyles(player.getEquipment().getWeaponId());
		int style = attackStyle;
		for (int i = style;i >= 0;i--) {
			if (styles.get(i) != null) {
				style = i;
				break;
			}
		}
		if (style != attackStyle)
			setAttackStyle(style);
		return styles.get(style);
	}

	public int getAttackBonusForStyle() {
		return getAttackStyle().getAttackType().getAttackBonus(player);
	}

	public int getDefenseBonusForStyle(AttackStyle style) {
		return style.getAttackType().getDefenseBonus(player);
	}

	public boolean isAutoRetaliate() {
		return autoRetaliate;
	}

	public void setAutoRetaliate(boolean autoRetaliate) {
		this.autoRetaliate = autoRetaliate;
	}

	public boolean isDungSpellbook() {
		return dungSpellbook;
	}

	public void removeDungeonneringBook() {
		if (dungSpellbook) {
			dungSpellbook = false;
			player.getInterfaceManager().sendSubDefault(Sub.TAB_MAGIC);
		}
	}

	public int getAttackStyleId() {
		return attackStyle;
	}
}

File: src/main/java/com/rs/game/content/combat/CombatDefinitions.kt
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
//  Copyright (C) 2021 Trenton Kress
//  This file is part of project: Darkan
//
package com.rs.game.content.combat

import com.rs.cache.loaders.Bonus
import com.rs.engine.quest.Quest
import com.rs.game.content.minigames.disableMinigameRunes
import com.rs.game.content.minigames.enableMinigameRunes
import com.rs.game.model.entity.player.Equipment
import com.rs.game.model.entity.player.Player
import com.rs.game.model.entity.player.managers.AuraManager
import com.rs.game.model.entity.player.managers.InterfaceManager
import com.rs.lib.util.Logger

class CombatDefinitions {
    enum class Spellbook(@JvmField val interfaceId: Int) {
        MODERN(192),
        ANCIENT(193),
        LUNAR(430),
        DUNGEONEERING(950)
    }

    @Transient
    private lateinit var player: Player

    @Transient
    var isUsingSpecialAttack: Boolean = false
        private set

    @Transient
    private lateinit var bonuses: IntArray

    @Transient
    var isDungSpellbook: Boolean = false
        private set

    private var attackStyle: Byte = 0
    private var specialAttackPercentage: Byte = 100
    var isAutoRetaliate: Boolean = true
    private var sortSpellBook: Byte = 0
    private var showCombatSpells = true
    private var showSkillSpells = true
    private var showMiscSpells = true
    private var showTeleportSpells = true
    var isDefensiveCasting: Boolean = false
        private set
    private var spellbook: Spellbook = Spellbook.MODERN
    var autoCast: CombatSpell? = null
        private set

    val spell: CombatSpell?
        get() {
            val spell = player.tempAttribs.getO<CombatSpell>("manualCastSpell")
            if (spell != null) return spell
            return autoCast
        }

    fun hasManualCastQueued(): Boolean {
        return player.tempAttribs.getO<Any?>("manualCastSpell") != null
    }

    fun setManualCastSpell(spell: CombatSpell?) {
        player.tempAttribs.setO<Any>("manualCastSpell", spell)
    }

    fun clearManualCastSpell() {
        player.tempAttribs.removeO<Any>("manualCastSpell")
    }

    fun resetSpells(removeAutoSpell: Boolean) {
        clearManualCastSpell()
        if (removeAutoSpell) {
            setAutoCastSpell(null)
            refreshAutoCastSpell()
        }
    }

    fun setAutoCastSpell(spell: CombatSpell?) {
        autoCast = spell
        refreshAutoCastSpell()
    }

    fun refreshAutoCastSpell() {
        refreshAttackStyle()
        player.vars.setVar(108, spellAutoCastConfigValue)
    }

    private val spellAutoCastConfigValue: Int
        get() {
            if (autoCast == null) return 0
            when {
                isDungSpellbook -> return when (autoCast) {
                    CombatSpell.WIND_STRIKE -> 103
                    CombatSpell.WATER_STRIKE -> 105
                    CombatSpell.EARTH_STRIKE -> 107
                    CombatSpell.FIRE_STRIKE -> 109
                    CombatSpell.WIND_BOLT -> 111
                    CombatSpell.WATER_BOLT -> 113
                    CombatSpell.EARTH_BOLT -> 115
                    CombatSpell.FIRE_BOLT -> 117
                    CombatSpell.WIND_BLAST -> 119
                    CombatSpell.WATER_BLAST -> 121
                    CombatSpell.EARTH_BLAST -> 123
                    CombatSpell.FIRE_BLAST -> 125
                    CombatSpell.WIND_WAVE -> 127
                    CombatSpell.WATER_WAVE -> 129
                    CombatSpell.EARTH_WAVE -> 131
                    CombatSpell.FIRE_WAVE -> 133
                    CombatSpell.WIND_SURGE -> 135
                    CombatSpell.WATER_SURGE -> 137
                    CombatSpell.EARTH_SURGE -> 139
                    CombatSpell.FIRE_SURGE -> 141
                    else -> 0
                }

                spellbook == Spellbook.MODERN -> return when (autoCast) {
                    CombatSpell.WIND_STRIKE -> 3
                    CombatSpell.WATER_STRIKE -> 5
                    CombatSpell.EARTH_STRIKE -> 7
                    CombatSpell.FIRE_STRIKE -> 9
                    CombatSpell.WIND_BOLT -> 11
                    CombatSpell.WATER_BOLT -> 13
                    CombatSpell.EARTH_BOLT -> 15
                    CombatSpell.FIRE_BOLT -> 17
                    CombatSpell.CRUMBLE_UNDEAD -> 35
                    CombatSpell.WIND_BLAST -> 19
                    CombatSpell.WATER_BLAST -> 21
                    CombatSpell.EARTH_BLAST -> 23
                    CombatSpell.FIRE_BLAST -> 25
                    CombatSpell.IBAN_BLAST -> 45
                    CombatSpell.MAGIC_DART -> 37
                    CombatSpell.SARADOMIN_STRIKE -> 41
                    CombatSpell.CLAWS_OF_GUTHIX -> 39
                    CombatSpell.FLAMES_OF_ZAMORAK -> 43
                    CombatSpell.WIND_WAVE -> 27
                    CombatSpell.WATER_WAVE -> 29
                    CombatSpell.EARTH_WAVE -> 31
                    CombatSpell.FIRE_WAVE -> 33
                    CombatSpell.WIND_SURGE -> 47
                    CombatSpell.WATER_SURGE -> 49
                    CombatSpell.EARTH_SURGE -> 51
                    CombatSpell.FIRE_SURGE -> 53
                    CombatSpell.WIND_RUSH -> 143
                    CombatSpell.STORM_OF_ARMADYL -> 145
                    else -> 0
                }

                spellbook == Spellbook.ANCIENT -> return when (autoCast) {
                    CombatSpell.SMOKE_RUSH -> 63
                    CombatSpell.SHADOW_RUSH -> 65
                    CombatSpell.BLOOD_RUSH -> 67
                    CombatSpell.ICE_RUSH -> 69
                    CombatSpell.SMOKE_BURST -> 71
                    CombatSpell.SHADOW_BURST -> 73
                    CombatSpell.BLOOD_BURST -> 75
                    CombatSpell.ICE_BURST -> 77
                    CombatSpell.SMOKE_BLITZ -> 79
                    CombatSpell.SHADOW_BLITZ -> 81
                    CombatSpell.BLOOD_BLITZ -> 83
                    CombatSpell.ICE_BLITZ -> 85
                    CombatSpell.SMOKE_BARRAGE -> 87
                    CombatSpell.SHADOW_BARRAGE -> 89
                    CombatSpell.BLOOD_BARRAGE -> 91
                    CombatSpell.ICE_BARRAGE -> 93
                    CombatSpell.MIASMIC_RUSH -> 95
                    CombatSpell.MIASMIC_BURST -> 97
                    CombatSpell.MIASMIC_BLITZ -> 99
                    CombatSpell.MIASMIC_BARRAGE -> 101
                    else -> 0
                }

            }
            return 0
        }

    fun setSpellbook(book: Spellbook) {
        if (book == Spellbook.LUNAR && !player.isQuestComplete(
                Quest.LUNAR_DIPLOMACY,
                "to use the Lunar spellbook."
            )
        ) return
        if (book == Spellbook.ANCIENT && !player.isQuestComplete(
                Quest.DESERT_TREASURE,
                "to use the Ancient spellbook."
            )
        ) return
        if (book == Spellbook.DUNGEONEERING) isDungSpellbook = true
        else spellbook = book
        refreshSpellbook()
        player.interfaceManager.sendSubDefault(InterfaceManager.Sub.TAB_MAGIC)
    }

    fun getSpellbook(): Spellbook? {
        if (isDungSpellbook) return Spellbook.DUNGEONEERING
        return spellbook
    }

    fun switchShowCombatSpells() {
        showCombatSpells = !showCombatSpells
        refreshSpellbookSettings()
    }

    fun switchShowSkillSpells() {
        showSkillSpells = !showSkillSpells
        refreshSpellbookSettings()
    }

    fun switchShowMiscSpells() {
        showMiscSpells = !showMiscSpells
        refreshSpellbookSettings()
    }

    fun switchShowTeleportSkillSpells() {
        showTeleportSpells = !showTeleportSpells
        refreshSpellbookSettings()
    }

    fun switchDefensiveCasting() {
        isDefensiveCasting = !isDefensiveCasting
        refreshSpellbookSettings()
    }

    fun setSortSpellBook(sortId: Int) {
        sortSpellBook = sortId.toByte()
        refreshSpellbookSettings()
    }

    fun refreshSpellbookSettings() {
        player.vars.setVarBit(357, spellbook.ordinal)
        player.vars.setVarBit(5822, sortSpellBook.toInt())
        player.vars.setVarBit(5823, sortSpellBook.toInt())
        player.vars.setVarBit(5824, sortSpellBook.toInt())
        player.vars.setVarBit(7347, sortSpellBook.toInt())

        player.vars.setVarBit(6459, if (showCombatSpells) 0 else 1)
        player.vars.setVarBit(6466, if (showCombatSpells) 0 else 1)
        player.vars.setVarBit(6463, if (showCombatSpells) 0 else 1)
        player.vars.setVarBit(7348, if (showCombatSpells) 0 else 1)

        player.vars.setVarBit(6460, if (showSkillSpells) 0 else 1)
        player.vars.setVarBit(7349, if (showSkillSpells) 0 else 1)

        player.vars.setVarBit(6461, if (showMiscSpells) 0 else 1)
        player.vars.setVarBit(6464, if (showMiscSpells) 0 else 1)
        player.vars.setVarBit(7350, if (showMiscSpells) 0 else 1)

        player.vars.setVarBit(6462, if (showTeleportSpells) 0 else 1)
        player.vars.setVarBit(6467, if (showTeleportSpells) 0 else 1)
        player.vars.setVarBit(6465, if (showTeleportSpells) 0 else 1)
        player.vars.setVarBit(7351, if (showTeleportSpells) 0 else 1)

        player.vars.setVarBit(2668, if (isDefensiveCasting) 1 else 0)
    }

    fun setPlayer(player: Player) {
        this.player = player
        bonuses = IntArray(18)
    }

    fun getBonus(bonus: Bonus): Int {
        return bonuses[bonus.ordinal]
    }

    fun setBonus(bonus: Bonus, `val`: Int) {
        bonuses[bonus.ordinal] = `val`
    }

    fun refreshBonuses() {
        bonuses = IntArray(18)
        for (item in player.equipment.itemsCopy) {
            if (item == null) continue
            for (bonus in Bonus.entries) {
                if (bonus == Bonus.RANGE_STR && getBonus(Bonus.RANGE_STR) != 0) continue
                bonuses[bonus.ordinal] += Equipment.getBonus(player, item, bonus)
            }
        }
    }

    fun resetSpecialAttack() {
        drainSpec(0)
        specialAttackPercentage = 100
        refreshSpecialAttackPercentage()
    }

    fun setSpecialAttack(special: Int) {
        drainSpec(0)
        specialAttackPercentage = special.toByte()
        refreshSpecialAttackPercentage()
    }

    fun restoreSpecialAttack() {
        if (player.familiar != null) player.familiar.restoreSpecialAttack(15)
        if (specialAttackPercentage.toInt() == 100) return
        var toRestore = 10
        if (player.auraManager.isActivated(AuraManager.Aura.INVIGORATE)) toRestore = 12
        else if (player.auraManager.isActivated(AuraManager.Aura.GREATER_INVIGORATE)) toRestore = 15
        else if (player.auraManager.isActivated(AuraManager.Aura.MASTER_INVIGORATE)) toRestore = 17
        else if (player.auraManager.isActivated(AuraManager.Aura.SUPREME_INVIGORATE)) toRestore = 20
        restoreSpecialAttack(toRestore)
        if (specialAttackPercentage.toInt() == 100 || specialAttackPercentage.toInt() == 50) player.sendMessage(
            "<col=00FF00>Your special attack energy is now $specialAttackPercentage%.",
            true
        )
    }

    fun restoreSpecialAttack(percentage: Int) {
        if (specialAttackPercentage >= 100 || player.interfaceManager.containsScreenInter()) return
        specialAttackPercentage =
            (specialAttackPercentage + if (specialAttackPercentage > (100 - percentage)) 100 - specialAttackPercentage else percentage).toByte()
        refreshSpecialAttackPercentage()
    }

    fun init() {
        refreshUsingSpecialAttack()
        refreshSpecialAttackPercentage()
        refreshAutoRelatie()
        refreshAttackStyle()
        refreshSpellbook()
    }

    fun refreshSpellbook() {
        refreshSpellbookSettings()
        refreshAutoCastSpell()
        if (isDungSpellbook)
            disableMinigameRunes(player)
        else
            enableMinigameRunes(player)
        player.vars.syncVarsToClient()
        player.packets.sendRunScriptBlank(2057)
    }

    fun checkAttackStyle() {
        if (autoCast == null) setAttackStyle(attackStyle.toInt())
    }

    fun setAttackStyle(style: Int) {
        var finalStyle = style
        val styles = AttackStyle.getStyles(
            player.equipment.weaponId
        )
        if (finalStyle < 0) finalStyle = 0
        for (i in finalStyle downTo 0) {
            if (styles[i] != null) {
                finalStyle = i
                break
            }
        }
        if (finalStyle != attackStyle.toInt()) {
            attackStyle = finalStyle.toByte()
            if (autoCast != null) resetSpells(true)
            else refreshAttackStyle()
        } else if (autoCast != null) resetSpells(true)
    }

    fun refreshAttackStyle() {
        player.vars.setVar(43, (if (autoCast != null) 4 else attackStyle).toInt())
    }

    fun sendUnlockAttackStylesButtons() {
        for (componentId in 7..10) player.packets.setIFRightClickOps(884, componentId, -1, 0, 0)
    }

    fun switchUsingSpecialAttack() {
        isUsingSpecialAttack = !isUsingSpecialAttack
        refreshUsingSpecialAttack()
    }

    fun drainSpec(amount: Int) {
        var finalAmount = amount
        isUsingSpecialAttack = false
        refreshUsingSpecialAttack()
        if (player.nsv.getB("infSpecialAttack")) finalAmount = 0
        if (finalAmount > 0) {
            specialAttackPercentage = (specialAttackPercentage - finalAmount).toByte()
            refreshSpecialAttackPercentage()
        }
    }

    fun hasRingOfVigour(): Boolean {
        return player.equipment.ringId == 19669
    }

    fun getSpecialAttackPercentage(): Int {
        return specialAttackPercentage.toInt()
    }

    fun refreshUsingSpecialAttack() {
        player.vars.setVar(301, if (isUsingSpecialAttack) 1 else 0)
    }

    fun refreshSpecialAttackPercentage() {
        player.vars.setVar(300, specialAttackPercentage * 10)
    }

    fun switchAutoRetaliate() {
        isAutoRetaliate = !isAutoRetaliate
        refreshAutoRelatie()
    }

    fun refreshAutoRelatie() {
        player.vars.setVar(172, if (isAutoRetaliate) 0 else 1)
    }

    fun getAttackStyle(): AttackStyle {
        val styles = AttackStyle.getStyles(
            player.equipment.weaponId
        )
        var styleIndex = attackStyle.toInt()
        for (i in styleIndex downTo 0) {
            if (styles[i] != null) {
                styleIndex = i
                break
            }
        }
        if (styleIndex != attackStyle.toInt()) setAttackStyle(styleIndex)
        val style = styles[styleIndex]
        if (style != null)
            return style
        else {
            Logger.handle(CombatDefinitions::class.java, "getAttackStyle", Error("Invalid attack style for weapon ${player.equipment.weaponId} on style index $styleIndex"))
            return AttackStyle(0, "Punch", XPType.ACCURATE, AttackType.CRUSH)
        }
    }

    val currentAttackBonus: Bonus
        get() = getAttackStyle().attackType.attBonus

    val currentDefenseBonus: Bonus
        get() = getAttackStyle().attackType.defBonus

    fun removeDungeonneringBook() {
        if (isDungSpellbook) {
            isDungSpellbook = false
            player.interfaceManager.sendSubDefault(InterfaceManager.Sub.TAB_MAGIC)
        }
    }

    val attackStyleId: Int
        get() = attackStyle.toInt()
}
